### ISSUE TITLE
ZMQ with Extism Plugins

### ISSUE THREAD

**Issue by neuronicnobody** [extism/extism#716]
If I wanted to use ZMQ (or another low level networking library) in a plugin how would I achieve this? What would be the proper way to include the header file and use the library functions provided. Attempting to compile ZMQ with an Extism C Plugin the following error is present:

```
terminate called after throwing an instance of 'extism::Error'
  what():  Unable to create Extism plugin: unknown import: `env::zmq_ctx_new` has not been defined
Aborted (core dumped)
```

Even if the networking calls are being proxied through the host application, the idea is to allow to different plugins running on different host machines to communicate via ZMQ without the host application developer having to interact with the low level details of the library

**Comment by bhelx** [comment]
is `env::zmq_ctx_new` a host function? where is this import coming from? if the module defines an import then you must satisfy it to load the plugin.

**Comment by HaoxinSEU** [comment]
Hi @bhelx ,

Thanks for the reply! `zmq_ctx_new` is a library function from `libzmq`. And I just realized that I should also build this library when I built my plugin. Now I'm following [this repository](https://github.com/mhmd-azeez/miniz_extism/tree/main) to include the ZMQ library in my `CMakeLists.txt`.

I changed to use `wasi-sdk/bin/wasm32-wasip2-clang` as it has `pthread`.

However, I met another error:
```
In file included from /opt/miniz_extism/dependencies/zmq/src/ws_address.cpp:3:
In file included from /opt/miniz_extism/dependencies/zmq/src/precompiled.hpp:30:
/opt/miniz_extism/dependencies/zmq/src/../include/zmq.h:769:33: error: unknown type name 'sigset_t'
  769 |                           const sigset_t *sigmask_);
      |                                 ^
In file included from /opt/miniz_extism/dependencies/zmq/src/ws_address.cpp:8:
In file included from /opt/miniz_extism/dependencies/zmq/src/ws_address.hpp:11:
/opt/miniz_extism/dependencies/zmq/src/ip_resolver.hpp:9:10: fatal error: 'netdb.h' file not found
    9 | #include <netdb.h>
      |          ^~~~~~~~~
2 errors generated.
make[2]: *** [CMakeFiles/objects.dir/build.make:76: CMakeFiles/objects.dir/src/ws_address.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:391: CMakeFiles/objects.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

It's very confusing because the header file `netdb.h` is actually in `wasi-sdk/share/wasi-sysroot/include/wasm32-wasip2/netdb.h`. 

Do you have any idea why this happens? Thanks again!

My `CMakeLists.txt` is like this:
```
cmake_minimum_required(VERSION 3.10)
project(miniz_extism)

# specify the C standard
set(CMAKE_C_STANDARD 99)

add_compile_options(-D_WASI_EMULATED_SIGNAL)

# specify the path to the Miniz library
set(MINIZ_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/miniz")
set(MINIZ_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/miniz")
add_subdirectory(${MINIZ_PATH} ${MINIZ_BINARY_DIR})
include_directories(${MINIZ_PATH})

# specify the path to the ZMQ library
set(ZMQ_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/zmq")
set(ZMQ_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/zmq")
add_subdirectory(${ZMQ_PATH} ${ZMQ_BINARY_DIR})
include_directories(${ZMQ_PATH}/include)

set(EXTISM_PATH "${CMAKE_CURRENT_SOURCE_DIR}/../dependencies/extism-pdk")
include_directories(${EXTISM_PATH})

# add the source file
add_executable(miniz_extism main.c)

# link the Miniz library
target_link_libraries(miniz_extism miniz libzmq wasi-emulated-signal)

set(CMAKE_C_FLAGS "--target=wasm32-wasip2")
set_target_properties(miniz_extism PROPERTIES OUTPUT_NAME miniz_extism.wasm)
```
And I run cmake with: `cmake -D CMAKE_C_COMPILER=$(pwd)/../../dependencies/extism-pdk/wasi-sdk/bin/wasm32-wasip2-clang -DCMAKE_LINKER=$(pwd)/../../dependencies/extism-pdk/wasi-sdk/bin/lld ..`

**Comment by G4Vi** [comment]
Some notes unrelated to cmake or the example, but with building with wasi-sdk:

> It's very confusing because the header file `netdb.h` is actually in `wasi-sdk/share/wasi-sysroot/include/wasm32-wasip2/netdb.h`.

I think when invoking clang you must specify the sysroot, for example `--sysroot=$WASI_SDK_PATH/share/wasi-sysroot`.

Extism isn't compatible with wasi preview 2 (`wasm32-wasip2`). With preview 1, I build with `"$WASI_SDK_PATH/bin/clang" "--sysroot=$WASI_SDK_PATH/share/wasi-sysroot" "--target=wasm32-wasi" "$@"`.

**Comment by HaoxinSEU** [comment]
Thanks @G4Vi !

I add this `sysroot` option in my `CMakeLists.txt`, but still get the same error.

I feel like `wasip2` is one of the reasons here. This `netdb.h` file is only available under `wasi-sysroot/include/wasm32-wasip2`, and is not available in preview 1 (`wasm32-wasi`, `wasm32-wasi-threads`, `wasm32-wasip1`, `wasm32-wasip1-threads`).

**Comment by bhelx** [comment]
Let us know how far you get, but I suspect you'll still have trouble compiling something that talks to sockets to wasip1. From my general understanding of zmq, it may be hard to compile and run it in wasm as it probably targets socket APIs at a low level (which wasip1 doesn't really support). You may be better off exposing what you need as [host functions](https://extism.org/docs/concepts/host-sdk) and running zmq on the host.

We are working up to having some support for wasip2 when we upgrade our version of wasmtime: https://github.com/extism/extism/pull/671

but keep in mind that wasip2 is only supported in wasmtime from what I know.

**Comment by HaoxinSEU** [comment]
I feel like it's a problem from WASI SDK. Because I tried to directly build `libzmq` with `wasm32-wasip2-clang`, but still got the same problem (cannot find `netdb.h`).  I opened an issue on wasi-sdk repository, maybe they have some insights.

Our idea is to run ZMQ communications in plugins, so that hosts are not bothered by this kind of low-level stuff. Putting them back to host functions seems to violate our intention....

Anyway, thanks a lot and we will keep you updated about this!
